### DISCUSSION TITLE
Typeclass AttrField (Like Django Model/Form Fields)

### DISCUSSION THREAD

**Discussion by strikaco** [evennia/evennia#2513] [Community Contribs & Snippets]
Typeclass attribute getter/setter shortcut class.

Defines type-specific attribute fields on Typeclasses, similar to how Django model/form fields are declared.

* A default value is a required kwarg. If dealing in setting objects as values, a null object instance must be set-- `None` as a value is not acceptable by default; string fields should have a blank string, numeric fields should have something akin to 0, etc.
* Maps to either the DBHandler, NDBHandler, or custom storage. You cannot use this to set properties on the model itself (prevents recursion errors and overwriting of things like `object.location` or `script.interval`).
* Schema is enforced globally but partitioned by storage, so you can't define `db.weight` as an `int` on one model and redefine it as a `str` on another. Provide `typed=False` as a kwarg to bypass this.

```
class AttrField(object):

    SCHEMA = {}

    DB = 'db'
    NDB = 'ndb'

    _MISSING = object()

    def __set_name__(self, instance, name):
        self._key = name

    def __init__(self, klass=None, *args, default, storage:str=DB, typed:bool=True, **kwargs):
        # Default value returned if not set
        self._default = default
        # Type of value accepted
        self._type = klass or type(default)
        # Whether or not to typecheck
        self._typed = typed
        # Where to get/set attributes from
        self._storage = storage

    def register_type(self, key, value):
        existing = self.SCHEMA.get(key, None)
        if not existing:
            self.SCHEMA[key] = value
            return

        if value != existing:
            raise TypeError(f'Key [{key}] associated with type {existing}; setting of {value} attempted.')

    def get_storage(self, instance):
        return getattr(instance, self._storage)
        
    def get_default(self):
        try:
            # Try to return a new instance of default if possible
            return self._default()
        except:
            # Return default instance itself
            return self._default

    def __set__(self, instance, value):
        # Validate type of value
        if self._typed:
            if value is None:
                value = self.get_default()
            else:
                self.register_type((self._storage, self._key), self._type)

        storage = self.get_storage(instance)
        setattr(storage, self._key, value)

    def __get__(self, instance, parent):
        if instance is None: return self
        
        storage = self.get_storage(instance)
        value = getattr(storage, self._key, None)
        if value is None:
            value = self.get_default()
            self.__set__(instance, value)
            return self.__get__(instance, parent)
        return value
```

Usage:

```
class Character(DefaultCharacter):
    title = AttrField(default='wanderer')
    desc = AttrField(default='a pasty white person')
    weight = AttrField(default=120.0)

char = Character.objects.first()
char.title = 'bored'
print(char.title)
print(char.db.title)
```

**Comment by Griatch** [comment]
It's a neat syntax! For some reason I've seen this style pop up in several suggestions all at once - in chat concerning composite fields from Kektus, this and I also added it as `TraitProperty` to the traits contrib the other day. 

The use of the `__get__(self, instance, parent)` magic method is very nice! I made use of the `@lazy_property` decorator for my implementation of `TraitProperty` (see below for an example), but I may not need that if using `__get__` under the hood instead. I should try that :grinning: 

```
from evennia.utils.utils import lazy_property
from evennia.contrib.traits import TraitProperty

class Character(DefaultCharacter):

    @lazy_property
    def strength(self):
        return TraitProperty(self, "Strength", trait_type='static', base=10)    
```

**Reply by strikaco** [reply]
Thanks!

I haven't seen the composite fields idea; I've just been irritated with the verbosity of `@property/@lazy_property` syntax...it takes 3 lines to define a single property, which makes it difficult to quickly see what fields/properties are expected in the class. Was going for presentation closer to Dataclasses or Django models.

The tradeoff here is customizability-- this way is not possible to configure per-instance.

**Reply by Griatch** [reply]
Well, even today, you'll modify your Attributes in the `at_object_creation` which is what makes it per-instance. 

I adopted the `__get__` mechanism for the `TraitProperties` now, very nice :+1: - thanks for the idea! 

> Edit: I see that @kektus also used `__get__` in this way for his components, I guess it just didn't register for me at the time :)

**Reply by strikaco** [reply]
Where can I find Kektus' version? I lost track of where to view the IRC logs and am curious what his solution is about.

I haven't seen your changes yet but be careful using `__get__`; if you're using it as part of a handler (instead of a proxy interface like mine) you'll want to return a new instance of the class instead of `self`. Otherwise unrelated objects will pick up on the reference to the previous object the handler was managing.

Ex:
```
class Handler(object):
    
    def __set_name__(self, instance, name):
        self._key = name
    
    def __init__(self, *args, **kwargs):
        self._args = args
        self._kwargs = kwargs
        self._obj = kwargs.pop('obj', None)
        self._key = kwargs.pop('key', None)
    
    def __get__(self, instance, parent):
        return self.__class__(*self._args, obj=instance, key=self._key, **self._kwargs)

    def handler_method(self, *args, **kwargs):
        self._obj.do_stuff()
```

**Reply by Griatch** [reply]
IRC is no more, unfortunately. You need to check in Discord, notably the ZFComponentStuff subthread. I can ask if Kektus'd be willing to make a post here as a community snippet.

Good point about the `__get__`. I'm not too familiar with descriptors. Will look my version over again, it's definitely possible there are some unexpected side effects like that.

**Reply by Griatch** [reply]
Went over my implementation of the TraitProperties again and fixed issues. Learned a lot about descriptors in the process of doing this :)
